### PR TITLE
Support Satyrographos

### PR DESCRIPTION
**Pull request by na4zagin3**
This PR makes `satysfi-matrix` installable with Satyrographos.

- Add `Satyristes` a build file.
- Add two OPAM package files `satysfi-matrix.opam` and `satysfi-matrix-doc.opam`.
- Update `README.md` and `README-ja.md`.
- Replace `@require: matrix` for `@require: matrix/matrix` in `example/main.saty`
  - This follows new Satyrographos directory hierarchy, which resembles the planned SATySFi package schema.

### CODE REVIEW

**Comment by nekketsuuu** [comment]
Thanks for your PR :) Since I personally haven't used Satyrographos ever, can I ask some questions?

* Is it OK to locate two `opam` files in the root directory? As far as I know, the OPAM documentation ([1][1], [2][2]) says "Package definitions can be a single `opam` file", and it doesn't say whether we can locate several `<pkgname>.opam` files. I think it may be strange to create two different package in one directory.
* Is the `Satyristes` file really necessary? It seems that that file contains what the opam file can contain.
* Is there any way to write the installation method in other than the `opam` file? If I want to publish `satysfi-matrix` to OPAM, then where to write the installation method without using Satyrographos? I don't want to officially use Satyrographos currently <del>because it is not published at <https://github.com/ocaml/opam></del>.


  [1]: https://opam.ocaml.org/doc/Packaging.html
  [2]: https://opam.ocaml.org/doc/Manual.html#Package-definitions

**Comment by nekketsuuu** [comment]
Oh, I'm sorry I'm wrong. Satyrographos has actually been released.
https://github.com/ocaml/opam-repository/tree/master/packages/satyrographos

**Comment by na4zagin3** [comment]
Of course I’m pleased to answer your questions! Probably, I should have asked for public review of the design of Satyrographos before sending MRs. The reason behind developing Satyrographos are:

- I want an abstract layer on top of SATySFi to manage differences of SATySFi versions.
- I believe network access and/or file system access needs to be separated from typesetting engines for security reason. (Nevertheless, I can decommission Satyrographos when SATySFi gets to have its own package management system after implementing migration support from Satyrographos libraries to SATySFi native libraries)

> Is it OK to locate two opam files in the root directory?

Yes. “Package definitions can be a single opam file” means does _not_ prohibit one repository having multiple packages.

1. In OPAM terminology, [“An opam package is defined by a <pkgname>.opam, or simply opam file, containing its metadata.”](https://opam.ocaml.org/doc/Packaging.html), in other words, a package is a bundle of distributed OCaml/Coq/etc. modules, resource files, and so on with a metadata described in a opam file.
1. OPAM commands can handle a repository with multiple packages. For example, `opam pin add [<package name>] <target>` can take `<package name>` as well as `<target>` which points a repository. OPAM Publish also supports such a repository and successfully submits a MR like https://github.com/na4zagin3/satyrographos-repo/pull/19.
1. There are many repositories containing multiple OPAM files: for example, [ocaml/dune](https://github.com/ocaml/dune), [andersfugmann/amqp-client](https://github.com/andersfugmann/amqp-client) and [ocaml/opam](https://github.com/ocaml/opam) itself.
1. That being said, your concern about having separated opam files for a package and its document is reasonable. OPAM supports build variables to enable/disable building and installing library documents. The reasons I decided to split them are following:
    1. Unlike other OCaml/Coq libraries, documents of a SATySFi package usually depends, by nature, on the SATySFi package during the build phase. The ideal package installation flow is 1) build the package (currently nothing happens here, though), 2) install the package, 3) build the doc, and 4) install the doc. But OPAM file does not have separated `build`/`install` sections for docs. Therefore I split them into two OPAM files (i.e., OPAM packages).
    1. I don’t want to make the problem complex although it is technically possible to install the packages somewhere else, probably under a temporary directory—wait, all the functionalities required for temporarily setting up SATYSFI_ROOT have been implemented. I’ll try that later. BTW, possibly can I have your opinion about this?

> Is the Satyristes file really necessary? It seems that that file contains what the opam file can contain.

Technically, no. But I believe providing a build file is better option.

1. With the build file, Satyrographos can place packages under correct locations even future SATySFi requires third party packages to follow the new schema (see https://github.com/na4zagin3/satyrographos/issues/24, https://satysfi.slack.com/archives/CBBFCSZ34/p1555826471028300).
1. Even without build files, you can technically install files where Satyrographos requires. For example, Satyrographos 0.0.1 required library authors to make proper Makefile like [satysfi-karnaugh](https://github.com/takagiy/satysfi-karnaugh/commit/b737f41701c5b9799d20bdeaafbb17263026e036#diff-b67911656ef5d18c4ae36cb6741b7965), but I don’t think it scales because Makefile can do anything, including locating files under wrong locations without getting any warnings. Providing a build file can solve that.
1. Satyrographos requires a metadata file to show compatibility warnings (e.g., “library _x_ has been renamed to _y_”)  to users of the library. To ensure metadata files’ validity, I could provide a command to validate metadata during installs. I chose another option where a valid metadata file is generated from the build file.

> It seems that that file contains what the opam file can contain.

Do you mean Satyristes file contain duplicated information or OPAM file can contain information written in Satyristes? You’re correct anyway, but I have reasons.

There are two reasons to have duplicated information. I want to automatically generate OPAM file from Satyristes file (though I haven’t finished implementation). Additionally I am planning to make Satyristes an independent package manager without runtime dependency on OPAM because of size of `~/.opam` directly (cf. https://github.com/amutake/satysfi-docker/issues/1).

It’s possible to add required information to OPAM file with extension fields `x-*` like the first option in https://github.com/na4zagin3/satyrographos/issues/13. I decided to invent a new format as it may be weird to use extension field if we want to cut off dependency on OPAM.

> Is there any way to write the installation method in other than the opam file? If I want to publish satysfi-matrix to OPAM, then where to write the installation method without using Satyrographos?

Could you elaborate that? Are you asking how to install without Satyrographos or OPAM or to publish a library to OPAM but without Satyrographos, or probably having a question of registering package to [satyrographos-repo](https://github.com/na4zagin3/satyrographos-repo) rather than [opam-repository](https://github.com/ocaml/opam-repository)?

Speaking of installation method without Satyrographos or OPAM, users can still manually copy `*.satyh` files under `SATYSFI_ROOT` because `satyrographos opam install` and `satyrographos install` simply copy files without modification (except for hash files being merged).

To have installation method with OPAM but without Satyrographos, there are two options:
1. Configure [filters](https://opam.ocaml.org/doc/Manual.html#Filters) in your OPAM file so that `satysfi install` is executed only when `satyrographos` is installed.
1. If you don’t want to have Satyrographos-related stuffs in your repositories, you have right to decline this MR, for sure.

For registering satyrographos-repo, I followed [Coq OPAM Repository](http://coq.io/opam/) which Coq libraries are registered to. OPAM repository is supposed to have OCaml libraries, therefore it’d better to have separated repositories for non-OCaml libraries.

**Comment by nekketsuuu** [comment]
Thank you for a long answer :)

> Unlike other OCaml/Coq libraries, documents of a SATySFi package usually depends, by nature, on the SATySFi package during the build phase.

Is [`with-doc`](https://opam.ocaml.org/doc/Manual.html#pkgvar-with-doc) filter not enough? We can use a local library to build a doc without installation (use `@import`). Also, since the doc of satysfi-matrix is written by Markdown, it's unnecessary for satysfi-matrix to separate builds.

>  I want to automatically generate OPAM file from Satyristes file.

It may be worth reading Dune's docs. Dune can recognize the `opam` file and also can generate the `opam` file automatically. It helps reducing the duplicate information.

> Could you elaborate that?

I asked that question because I don't want to depend only on Satyrographos to install satysfi-matrix, for now. For example, if I write the `opam` file, I will write simple copy commands to `install:`.

Locating the `Satyristes` file is really needed? The OPAM repository permits to locate setting files in OPAM's package repository and not in package's source repository.

Importantly, I cannot maintain this package frequently because this is a toy library. So I've not considered to publish this package ever; it's dangerous. I will not be able to maintain the Satyrographos repository.

In summary, my main concern is: Is it really needed to add files here to publish a package to Satyrographos?

**Comment by na4zagin3** [comment]
@nekketsuuu Sorry for my late reply. Sure! Till you change your mind I maintain packaging like https://github.com/na4zagin3/satyrographos-repo/pull/29. We’ll see.

I’d like to thank you for publishing your useful libraries. Especially satysfi-matrix is worth to distribute as a Satyrographos package. Let me know if you need anything. I would be glad if I could have helped you in any way.
